Import the imdb.csv data file into a Pandas DataFrame and name it **df**

In [2]:
import pandas as pd
df = pd.read_csv('imdb.csv')
df

Rank                    Title                     Genre  \
0       1  Guardians of the Galaxy   Action,Adventure,Sci-Fi   
1       2               Prometheus  Adventure,Mystery,Sci-Fi   
2       3                    Split           Horror,Thriller   
3       4                     Sing   Animation,Comedy,Family   
4       5            Suicide Squad  Action,Adventure,Fantasy   
..    ...                      ...                       ...   
995   996     Secret in Their Eyes       Crime,Drama,Mystery   
996   997          Hostel: Part II                    Horror   
997   998   Step Up 2: The Streets       Drama,Music,Romance   
998   999             Search Party          Adventure,Comedy   
999  1000               Nine Lives     Comedy,Family,Fantasy   

                                           Description              Director  \
0    A group of intergalactic criminals are forced ...            James Gunn   
1    Following clues to the origin of mankind, a te...          Ridley Scott   
2    Three girls are kidnapped by a man with a diag...    M. Night Shyamalan   
3    In a city of humanoid animals, a hustling thea...  Christophe Lourdelet   
4    A secret government agency recruits some of th...            David Ayer   
..                                                 ...                   ...   
995  A tight-knit team of rising investigators, alo...             Billy Ray   
996  Three American college students studying abroa...              Eli Roth   
997  Romantic sparks occur between two dance studen...            Jon M. Chu   
998  A pair of friends embark on a mission to reuni...        Scot Armstrong   
999  A stuffy businessman finds himself trapped ins...      Barry Sonnenfeld   

                                                Actors  Year  \
0    Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...  2014   
1    Noomi Rapace, Logan Marshall-Green, Michael Fa...  2012   
2    James McAvoy, Anya Taylor-Joy, Haley Lu Richar...  2016   
3    Matthew McConaughey,Reese Witherspoon, Seth Ma...  2016   
4    Will Smith, Jared Leto, Margot Robbie, Viola D...  2016   
..                                                 ...   ...   
995  Chiwetel Ejiofor, Nicole Kidman, Julia Roberts...  2015   
996  Lauren German, Heather Matarazzo, Bijou Philli...  2007   
997  Robert Hoffman, Briana Evigan, Cassie Ventura,...  2008   
998  Adam Pally, T.J. Miller, Thomas Middleditch,Sh...  2014   
999  Kevin Spacey, Jennifer Garner, Robbie Amell,Ch...  2016   

     Runtime (Minutes)  Rating   Votes  Revenue (Millions)  Metascore  
0                  121     8.1  757074              333.13       76.0  
1                  124     7.0  485820              126.46       65.0  
2                  117     7.3  157606              138.12       62.0  
3                  108     7.2   60545              270.32       59.0  
4                  123     6.2  393727              325.02       40.0  
..                 ...     ...     ...                 ...        ...  
995                111     6.2   27585                 NaN       45.0  
996                 94     5.5   73152               17.54       46.0  
997                 98     6.2   70699               58.01       50.0  
998                 93     5.6    4881                 NaN       22.0  
999                 87     5.3   12435               19.64       11.0  

[1000 rows x 12 columns]

The code in the code cell below generates a unique list of genres based on each of those that appear in the Genre column. 

In [3]:
genres = []

for row in df.Genre:            # for each row in the Genre column
  for genre in row.split(','):  # for each genre in the row
    if genre not in genres:     # if that genre is not already in the unique genre list
      genres.append(genre)      # add it to the list
      
genres

['Action',
 'Adventure',
 'Sci-Fi',
 'Mystery',
 'Horror',
 'Thriller',
 'Animation',
 'Comedy',
 'Family',
 'Fantasy',
 'Drama',
 'Music',
 'Biography',
 'Romance',
 'History',
 'Crime',
 'Western',
 'War',
 'Musical',
 'Sport']

Notice that many of the genres in the original table have more than one value. This is an issue. Therefore:
* Create a new column in the imdb.csv DataFrame for each of those genres. Use False as the default value. HINT: if you use a loop to iterate through the genre list, you can accomplish this in only two, simple lines of code
* Iterate through the rows of the DataFrame and update the values from False to True for each of the appropriate genre columns if that genre appears in the original Genre column

For example, if a particular row has a value of 'Drama, Comedy' for the Genre column, the following columns should give the values:
* Action = False
* Adventure = False
* Sci-Fi = False
* Mystery = False
* Horror = False
* Thriller = False
* Animation = False
* Comedy = True
* Family = False
* Fantasy = False
* Drama = True
* Music = False
* Biography = False
* Romance = False
* History = False
* Crime = False
* Western = False
* War = False
* Musical = False
* Sport = False

Print out the new dataframe

In [4]:
# Question 2:

for genre in genres:
  df[genre] = False
  
for row in df.itertuples():       # for each row in the DataFrame
  for col in genres:              # for each genre in the unique genres list
    if col in row.Genre:          # if the original Genres column contains that genre
      df.loc[row[0], col] = True  # update the cell value to 1; row[0] refers to the Index of the row object in the first, outer loop
      
df

Rank                    Title                     Genre  \
0       1  Guardians of the Galaxy   Action,Adventure,Sci-Fi   
1       2               Prometheus  Adventure,Mystery,Sci-Fi   
2       3                    Split           Horror,Thriller   
3       4                     Sing   Animation,Comedy,Family   
4       5            Suicide Squad  Action,Adventure,Fantasy   
..    ...                      ...                       ...   
995   996     Secret in Their Eyes       Crime,Drama,Mystery   
996   997          Hostel: Part II                    Horror   
997   998   Step Up 2: The Streets       Drama,Music,Romance   
998   999             Search Party          Adventure,Comedy   
999  1000               Nine Lives     Comedy,Family,Fantasy   

                                           Description              Director  \
0    A group of intergalactic criminals are forced ...            James Gunn   
1    Following clues to the origin of mankind, a te...          Ridley Scott   
2    Three girls are kidnapped by a man with a diag...    M. Night Shyamalan   
3    In a city of humanoid animals, a hustling thea...  Christophe Lourdelet   
4    A secret government agency recruits some of th...            David Ayer   
..                                                 ...                   ...   
995  A tight-knit team of rising investigators, alo...             Billy Ray   
996  Three American college students studying abroa...              Eli Roth   
997  Romantic sparks occur between two dance studen...            Jon M. Chu   
998  A pair of friends embark on a mission to reuni...        Scot Armstrong   
999  A stuffy businessman finds himself trapped ins...      Barry Sonnenfeld   

                                                Actors  Year  \
0    Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...  2014   
1    Noomi Rapace, Logan Marshall-Green, Michael Fa...  2012   
2    James McAvoy, Anya Taylor-Joy, Haley Lu Richar...  2016   
3    Matthew McConaughey,Reese Witherspoon, Seth Ma...  2016   
4    Will Smith, Jared Leto, Margot Robbie, Viola D...  2016   
..                                                 ...   ...   
995  Chiwetel Ejiofor, Nicole Kidman, Julia Roberts...  2015   
996  Lauren German, Heather Matarazzo, Bijou Philli...  2007   
997  Robert Hoffman, Briana Evigan, Cassie Ventura,...  2008   
998  Adam Pally, T.J. Miller, Thomas Middleditch,Sh...  2014   
999  Kevin Spacey, Jennifer Garner, Robbie Amell,Ch...  2016   

     Runtime (Minutes)  Rating   Votes  ...  Drama  Music  Biography  Romance  \
0                  121     8.1  757074  ...  False  False      False    False   
1                  124     7.0  485820  ...  False  False      False    False   
2                  117     7.3  157606  ...  False  False      False    False   
3                  108     7.2   60545  ...  False  False      False    False   
4                  123     6.2  393727  ...  False  False      False    False   
..                 ...     ...     ...  ...    ...    ...        ...      ...   
995                111     6.2   27585  ...   True  False      False    False   
996                 94     5.5   73152  ...  False  False      False    False   
997                 98     6.2   70699  ...   True   True      False     True   
998                 93     5.6    4881  ...  False  False      False    False   
999                 87     5.3   12435  ...  False  False      False    False   

     History  Crime  Western    War  Musical  Sport  
0      False  False    False  False    False  False  
1      False  False    False  False    False  False  
2      False  False    False  False    False  False  
3      False  False    False  False    False  False  
4      False  False    False  False    False  False  
..       ...    ...      ...    ...      ...    ...  
995    False   True    False  False    False  False  
996    False  False    False  False    False  False  
997    False  False    False  False    False  F

In [5]:
#drop missing values in order to fix potential issues (we will discuss this later in the course)
df = df.dropna()
df

Rank                     Title                     Genre  \
0       1   Guardians of the Galaxy   Action,Adventure,Sci-Fi   
1       2                Prometheus  Adventure,Mystery,Sci-Fi   
2       3                     Split           Horror,Thriller   
3       4                      Sing   Animation,Comedy,Family   
4       5             Suicide Squad  Action,Adventure,Fantasy   
..    ...                       ...                       ...   
993   994  Resident Evil: Afterlife   Action,Adventure,Horror   
994   995                 Project X                    Comedy   
996   997           Hostel: Part II                    Horror   
997   998    Step Up 2: The Streets       Drama,Music,Romance   
999  1000                Nine Lives     Comedy,Family,Fantasy   

                                           Description              Director  \
0    A group of intergalactic criminals are forced ...            James Gunn   
1    Following clues to the origin of mankind, a te...          Ridley Scott   
2    Three girls are kidnapped by a man with a diag...    M. Night Shyamalan   
3    In a city of humanoid animals, a hustling thea...  Christophe Lourdelet   
4    A secret government agency recruits some of th...            David Ayer   
..                                                 ...                   ...   
993  While still out to destroy the evil Umbrella C...    Paul W.S. Anderson   
994  3 high school seniors throw a birthday party t...       Nima Nourizadeh   
996  Three American college students studying abroa...              Eli Roth   
997  Romantic sparks occur between two dance studen...            Jon M. Chu   
999  A stuffy businessman finds himself trapped ins...      Barry Sonnenfeld   

                                                Actors  Year  \
0    Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...  2014   
1    Noomi Rapace, Logan Marshall-Green, Michael Fa...  2012   
2    James McAvoy, Anya Taylor-Joy, Haley Lu Richar...  2016   
3    Matthew McConaughey,Reese Witherspoon, Seth Ma...  2016   
4    Will Smith, Jared Leto, Margot Robbie, Viola D...  2016   
..                                                 ...   ...   
993  Milla Jovovich, Ali Larter, Wentworth Miller,K...  2010   
994  Thomas Mann, Oliver Cooper, Jonathan Daniel Br...  2012   
996  Lauren German, Heather Matarazzo, Bijou Philli...  2007   
997  Robert Hoffman, Briana Evigan, Cassie Ventura,...  2008   
999  Kevin Spacey, Jennifer Garner, Robbie Amell,Ch...  2016   

     Runtime (Minutes)  Rating   Votes  ...  Drama  Music  Biography  Romance  \
0                  121     8.1  757074  ...  False  False      False    False   
1                  124     7.0  485820  ...  False  False      False    False   
2                  117     7.3  157606  ...  False  False      False    False   
3                  108     7.2   60545  ...  False  False      False    False   
4                  123     6.2  393727  ...  False  False      False    False   
..                 ...     ...     ...  ...    ...    ...        ...      ...   
993                 97     5.9  140900  ...  False  False      False    False   
994                 88     6.7  164088  ...  False  False      False    False   
996                 94     5.5   73152  ...  False  False      False    False   
997                 98     6.2   70699  ...   True   True      False     True   
999                 87     5.3   12435  ...  False  False      False    False   

     History  Crime  Western    War  Musical  Sport  
0      False  False    False  False    False  False  
1      False  False    False  False    False  False  
2      False  False    False  False    False  False  
3      False  False    False  False    False  False  
4      False  False    False  False    False  False  
..       ...    ...      ...    ...      ...    ...  
993    False  False    False  False    False  False  
994    False  False    False  False    False  False  
996    False  False    False  False

**NEXT**: Let's analyze the effect of genre on Rating. Create 5 lists of Ratings by **filling in the following code**--one list for each of the following genres: Action, Drama, Comedy, Horror, Romance

Compare these lists of ratings in an ANOVA to see if there are any significant differences. Print out the F and p-value from this ANOVA.

In [6]:
from scipy import stats
import statistics as stat
action = df[df['Action'] == True]['Rating']
drama = df[df['Drama'] == True]['Rating']
comedy = df[df['Comedy'] == True]['Rating']
horror = df[df['Horror'] == True]['Rating']
romance = df[df['Romance'] == True]['Rating']
F, p = stats.f_oneway(action, drama, comedy, horror, romance)


print("F-value:", F)
print("p-value:", p)


F-value: 20.29166363479183
p-value: 3.631197919868067e-16


**ANSWER**: How do you interpret the results? Be very specific, and write your answers in this markdown section.

**Next**: Print the means of each of the genre ratings.

The results from the ANOVA test show that the F-value is 20.29, and the p-value is extremely small, about 3.63e-16. This suggests there are significant statistical differences in the average ratings between the different genres: Action, Drama, Comedy, Horror, and Romance. Since the p-value is so small, it indicates that it’s highly unlikely these differences happened by chance. In other words, the genre of a movie seems to have a noticeable impact on its average rating. Currently, we do not do which genre in specific are driving these differences, but overall, it's clear that not all genres have the same average ratings.

In [7]:
# Calculate and print the mean ratings for each genre
mean_action = df[df['Action'] == True]['Rating'].mean()
mean_drama = df[df['Drama'] == True]['Rating'].mean()
mean_comedy = df[df['Comedy'] == True]['Rating'].mean()
mean_horror = df[df['Horror'] == True]['Rating'].mean()
mean_romance = df[df['Romance'] == True]['Rating'].mean()

print("Mean Ratings:")
print("Action:", mean_action)
print("Drama:", mean_drama)
print("Comedy:", mean_comedy)
print("Horror:", mean_horror)
print("Romance:", mean_romance)



Mean Ratings:
Action: 6.650902527075812
Drama: 7.042959427207637
Comedy: 6.7092
Horror: 6.288505747126436
Romance: 6.700833333333333


**ANSWER**: Based on your ANOVA results, how can you interpret the means? Again, make sure to communicate something meaningful to interpret the data, not just numbers.

The ANOVA test results show that the differences in average ratings between the movie genres are statistically significant, with an F-value of 20.29 and an extremely small p-value of about 3.63e-16. This means that it's very unlikely these differences happened by chance, so the genre of a movie clearly influences its average rating.

When looking at the average ratings, Drama has the highest average at 7.04, while Horror has the lowest at 6.29. The difference between these two genres is especially noticeable. This sharp contrast suggests that viewers tend to rate dramas much higher than horror movies. With Drama being rated much higher than Horror, it’s likely that the significant differences highlighted by the ANOVA test are largely driven by this contrast. The other genres, like Action, Comedy, and Romance, fall somewhere in the middle with average ratings close to 6.65 to 6.71, but the biggest gap is clearly between Drama and Horror, which helps explain why the overall differences are statistically significant. So, it’s clear that people’s ratings vary by genre, with some genres like Drama generally getting better reviews than others like Horror.

**NEXT**: Let's see if the critics agree with the movie-goers. **Perform the same analyses using Metascore** (a score generated by critics) as the label. Compare the same five genres and print out both their mean Metascores. **Comment on your results**

In [10]:
from scipy import stats

action_metascore = df[df['Action'] == True]['Metascore']
drama_metascore = df[df['Drama'] == True]['Metascore']
comedy_metascore = df[df['Comedy'] == True]['Metascore']
horror_metascore = df[df['Horror'] == True]['Metascore']
romance_metascore = df[df['Romance'] == True]['Metascore']


F_metascore, p_metascore = stats.f_oneway(action_metascore, drama_metascore, comedy_metascore, horror_metascore, romance_metascore)

mean_action_metascore = action_metascore.mean()
mean_drama_metascore = drama_metascore.mean()
mean_comedy_metascore = comedy_metascore.mean()
mean_horror_metascore = horror_metascore.mean()
mean_romance_metascore = romance_metascore.mean()


print("Mean Metascores:")
print("Action:", mean_action_metascore)
print("Drama:", mean_drama_metascore)
print("Comedy:", mean_comedy_metascore)
print("Horror:", mean_horror_metascore)
print("Romance:", mean_romance_metascore)

print("\nANOVA Results:")
print("F-value:", F_metascore)
print("p-value:", p_metascore)


Mean Metascores:
Action: 54.04693140794224
Drama: 64.57756563245823
Comedy: 58.28
Horror: 56.45977011494253
Romance: 55.46666666666667

ANOVA Results:
F-value: 20.243215736410303
p-value: 3.966530651621933e-16


The ANOVA test results for Metascores show that critics rate movies differently across the five genres, with an F-value of 20.24 and a super low p-value of 3.97e-16. This means the differences in average Metascores are statistically significant. Drama really stands out as a clear outlier, with a mean Metascore of 64.58, which suggests that critics generally like dramas more than the other genres. On the other hand, Action has the lowest mean Metascore at 54.05, meaning critics aren’t as impressed with action films. Comedy (58.28), Horror (56.46), and Romance (55.47) have mean Metascores that are pretty close but still fall far below Drama.

These results highlight a big contrast in how critics see the quality of films based on genre. The noticeable gap between Drama and Action, along with the other genres, shows that Drama is rated much higher than the rest. This difference suggests that critics really favor dramas, while audience ratings might not always match this view.

**NEXT**: Let's run one more analyses on the Year of the movie. Even though Year is numeric and has an order, we do not have a theory (i.e. a logical reason) to suspect that movies get better or worse over time. Therefore, we should **analyze Year as a nominal/object** data type using an ANOVA. 

Therefore, convert the Year column to an object and then compare the Rating of the years 2012-2016 by **completing the following code** to print out both the mean Rating for each of those years and the F and p-value of an ANOVA comparing them. **Comment on your analysis**

In [12]:

label = 'Rating'

# Extract ratings for each year
y2012 = df[df['Year'] == '2012'][label]
y2013 = df[df['Year'] == '2013'][label]
y2014 = df[df['Year'] == '2014'][label]
y2015 = df[df['Year'] == '2015'][label]
y2016 = df[df['Year'] == '2016'][label]

F_year, p_year = stats.f_oneway(y2012, y2013, y2014, y2015, y2016)

mean_y2012 = y2012.mean()
mean_y2013 = y2013.mean()
mean_y2014 = y2014.mean()
mean_y2015 = y2015.mean()
mean_y2016 = y2016.mean()

print("Mean Ratings:")
print("2012:", mean_y2012)
print("2013:", mean_y2013)
print("2014:", mean_y2014)
print("2015:", mean_y2015)
print("2016:", mean_y2016)

print("\nANOVA Results:")
print("F-value:", F_year)
print("p-value:", p_year)

Mean Ratings:
2012: 6.933870967741935
2013: 6.832142857142857
2014: 6.82258064516129
2015: 6.674311926605503
2016: 6.644444444444444

ANOVA Results:
F-value: 2.105123928698873
p-value: 0.07889378412064842


The ANOVA test results reveal an F-value of 2.11 and a p-value of 0.079. Since the p-value is slightly above the conventional threshold of 0.05, we cannot conclude that there are statistically significant differences in average movie ratings across the years 2012 to 2016. When examining the mean ratings, 2012 stands out with the highest average at 6.93, while 2016 has the lowest at 6.64. Although these differences suggest a slight decline in ratings over the years, the lack of statistical significance indicates that these variations may not be substantial enough to draw strong conclusions. Overall, while the data shows some variation in ratings year by year, it doesn't provide robust evidence that ratings significantly differ based on the year alone. Perhaps there might be other factors to suggest this idea.